pkg/ipfs/service.go

```diff
@@ -0,0 +1,411 @@
+	Publisher IPLDPublisher
+	// Interface for indexing the CIDs of the published ETH-IPLDs in Postgres
+	Repository CIDRepository
+	// Interface for filtering and serving data according to subscribed clients according to their specification
+	Screener ResponseScreener
+	// Interface for fetching ETH-IPLD objects from IPFS
+	Fetcher IPLDFetcher
+	// Interface for searching and retrieving CIDs from Postgres index
+	Retriever CIDRetriever
+	// Interface for resolving ipfs blocks to their data types
+	Resolver IPLDResolver
+	// Chan the processor uses to subscribe to state diff payloads from the Streamer
+	PayloadChan chan statediff.Payload
+	// Used to signal shutdown of the service
+	QuitChan chan bool
+	// A mapping of rpc.IDs to their subscription channels, mapped to their subscription type (hash of the StreamFilters)
+	Subscriptions map[common.Hash]map[rpc.ID]Subscription
+	// A mapping of subscription hash type to the corresponding StreamFilters
+	SubscriptionTypes map[common.Hash]config.Subscription
+}
+
+// NewIPFSProcessor creates a new Processor interface using an underlying Processor struct
+func NewIPFSProcessor(ipfsPath string, db *postgres.DB, ethClient core.EthClient, rpcClient core.RpcClient, qc chan bool) (SyncPublishScreenAndServe, error) {
+	publisher, err := NewIPLDPublisher(ipfsPath)
+	if err != nil {
+		return nil, err
+	}
+	fetcher, err := NewIPLDFetcher(ipfsPath)
+	if err != nil {
+		return nil, err
+	}
+	return &Service{
+		Streamer:          NewStateDiffStreamer(rpcClient),
+		Repository:        NewCIDRepository(db),
+		Converter:         NewPayloadConverter(ethClient),
+		Publisher:         publisher,
+		Screener:          NewResponseScreener(),
+		Fetcher:           fetcher,
+		Retriever:         NewCIDRetriever(db),
+		Resolver:          NewIPLDResolver(),
+		PayloadChan:       make(chan statediff.Payload, payloadChanBufferSize),
+		QuitChan:          qc,
+		Subscriptions:     make(map[common.Hash]map[rpc.ID]Subscription),
+		SubscriptionTypes: make(map[common.Hash]config.Subscription),
+	}, nil
+}
+
+// Protocols exports the services p2p protocols, this service has none
+func (sap *Service) Protocols() []p2p.Protocol {
+	return []p2p.Protocol{}
+}
+
+// APIs returns the RPC descriptors the StateDiffingService offers
+func (sap *Service) APIs() []rpc.API {
+	return []rpc.API{
+		{
+			Namespace: APIName,
+			Version:   APIVersion,
+			Service:   NewPublicSeedNodeAPI(sap),
+			Public:    true,
+		},
+	}
+}
+
+// SyncAndPublish is the backend processing loop which streams data from geth, converts it to iplds, publishes them to ipfs, and indexes their cids
+// This continues on no matter if or how many subscribers there are, it then forwards the data to the ScreenAndServe() loop
+// which filters and sends relevant data to client subscriptions, if there are any
+func (sap *Service) SyncAndPublish(wg *sync.WaitGroup, screenAndServePayload chan<- IPLDPayload, screenAndServeQuit chan<- bool) error {
```

Comment:
I'm finding it a bit confusing that a method called `SyncAndPublish` takes in channels called `screenAndServePayload` even though it's possible that they won't be used (especially since they're not used in the `syncAndPublish` command). 

I understand why it's structured this way though, so that we can publish the payloads in several workers, and then pass the converted payload along to serve it. But I wonder if these two methods are responsible for too many things. 🤔

Reply:
Just playing around, not sure if this is much better. But what about replaces `SyncAndPublish` with:

```go
func (sap *Service) Convert(wg *sync.WaitGroup, convertedPayloadsCh chan<- IPLDPayload, convertedPayloadsQuitCh chan<- bool) error {
	sub, err := sap.Streamer.Stream(sap.PayloadChan)
	if err != nil {
		return err
	}
	wg.Add(1)

	go func() {
		for {
			select {
			case payload := <-sap.PayloadChan:
				if payload.Err != nil {
					log.Error(err)
					continue
				}
				ipldPayload, err := sap.Converter.Convert(payload)
				if err != nil {
					log.Error(err)
					continue
				}

				convertedPayloadsCh <- *ipldPayload

			case err = <-sub.Err():
				log.Error(err)
			case <-sap.QuitChan:
				// If we have a ScreenAndServe process running, forward the quit signal to it
				case convertedPayloadsQuitCh <- true:
				// Also forward a quit signal for each of the workers
				log.Info("quiting SyncAndPublish process")
				wg.Done()
				return
			}
		}
	}()

	return nil
}

func (sap *Service) DelegateIPLDPayloads(wg *sync.WaitGroup, convertedPayloadsCh chan IPLDPayload, convertedPayloadsQuitCh chan bool) error {
	// Channels for forwarding data to the publishAndIndex workers
	publishAndIndexPayload := make(chan IPLDPayload, payloadChanBufferSize)
	publishAndIndexQuit := make(chan bool, workerPoolSize)
	// publishAndIndex worker pool to handle publishing and indexing concurrently, while
	// limiting the number of Postgres connections we can possibly open so as to prevent error
	for i := 0; i < workerPoolSize; i++ {
		sap.publishAndIndex(i, publishAndIndexPayload, publishAndIndexQuit)
	}

	ScreenAndServeOn := true
	var screenAndServePayload chan IPLDPayload
	var screenAndServeQuit chan bool
	
    if  ScreenAndServeOn {
		screenAndServePayload := make(chan IPLDPayload, payloadChanBufferSize)
		screenAndServeQuit := make(chan bool, workerPoolSize)
		sap.ScreenAndServe(screenAndServePayload, screenAndServeQuit)
	}

	go func() {
		for {
			select {
			case ipldPayload := <-convertedPayloadsCh:
			    publishAndIndexPayload <- ipldPayload
				screenAndServePayload <- ipldPayload
			case <-sap.QuitChan:
				publishAndIndexPayload <- true
				if ScreenAndServeOn {
					screenAndServeQuit <- true
				}
				log.Info("quiting SyncAndPublish process")
				wg.Done()
				return
			}
		}
	}()
}
```

Reply:
The idea was that the `SyncAndPublish` does all the sync and publishing, and that it can optionally be asynchronously paired with the "ScreenAndServe" method using the provided `screenAndServePayload` channel. I need to take a closer look at this replacement code, I think the new naming makes sense but I think there are few issues with the specifics. e.g.`ScreenAndServeOn` being hard coded as `true` and the need for a `ScreenAndServeOn` global variable in the first place- previous code uses a `select` to send to the `screenAndServePayload` channel if it is non-nil, otherwise it just `default`s past it; it doesn't care/need to check if the ScreenAndServe process is also running. In any case the sends to the `screenAndServePayload` and `screenAndServeQuit` should be through `select` statements so that we don't get suck if the channel is full.

Reply:
I don't know where @rmulhol is in this process. But we probably want to take a more functional approach to the design of vDb and create new more cleanly composable functions:
`sync`, `publish`, `filter`, and `serve`. I don't understand why we'd use two verbs in one function.

Reply:
I agree! Although I need to think about how this would be implemented a bit more. My original line of thinking was that we can't publish without syncing, and there's no need/we don't know in what way to filter the data unless we are serving a subscription.